drivers/osdp/src/osdp_pd.c

```diff
@@ -0,0 +1,576 @@
+		cmd->led.temporary.off_count    = buf[pos++];
+		cmd->led.temporary.on_color     = buf[pos++];
+		cmd->led.temporary.off_color    = buf[pos++];
+		cmd->led.temporary.timer        = buf[pos++];
+		cmd->led.temporary.timer       |= buf[pos++] << 8;
+
+		cmd->led.permanent.control_code = buf[pos++];
+		cmd->led.permanent.on_count     = buf[pos++];
+		cmd->led.permanent.off_count    = buf[pos++];
+		cmd->led.permanent.on_color     = buf[pos++];
+		cmd->led.permanent.off_color    = buf[pos++];
+		pd_enqueue_command(pd, cmd);
+		reply->id = REPLY_ACK;
+		ret = 0;
+		break;
+	case CMD_BUZ:
+		if (len != 5)
+			break;
+		cmd = osdp_cmd_alloc();
+		if (cmd == NULL) {
+			LOG_ERR(TAG "cmd alloc error");
+			break;
+		}
+		cmd->id = OSDP_CMD_BUZZER;
+		cmd->buzzer.reader    = buf[pos++];
+		cmd->buzzer.tone_code = buf[pos++];
+		cmd->buzzer.on_count  = buf[pos++];
+		cmd->buzzer.off_count = buf[pos++];
+		cmd->buzzer.rep_count = buf[pos++];
+		pd_enqueue_command(pd, cmd);
+		reply->id = REPLY_ACK;
+		ret = 0;
+		break;
+	case CMD_TEXT:
+		if (len < 7)
+			break;
+		cmd = osdp_cmd_alloc();
+		if (cmd == NULL) {
+			LOG_ERR(TAG "cmd alloc error");
+			break;
+		}
+		cmd->id = OSDP_CMD_TEXT;
+		cmd->text.reader     = buf[pos++];
+		cmd->text.cmd        = buf[pos++];
+		cmd->text.temp_time  = buf[pos++];
+		cmd->text.offset_row = buf[pos++];
+		cmd->text.offset_col = buf[pos++];
+		cmd->text.length     = buf[pos++];
+		if (cmd->text.length > 32)
+			break;
+		for (i = 0; i < cmd->text.length; i++)
+			cmd->text.data[i] = buf[pos++];
+		pd_enqueue_command(pd, cmd);
+		reply->id = REPLY_ACK;
+		ret = 0;
+		break;
+	case CMD_COMSET:
+		if (len != 5)
+			break;
+		cmd = osdp_cmd_alloc();
+		if (cmd == NULL) {
+			LOG_ERR(TAG "cmd alloc error");
+			break;
+		}
+		cmd->id = OSDP_CMD_COMSET;
+		cmd->comset.addr  = buf[pos++];
+		cmd->comset.baud  = buf[pos++];
+		cmd->comset.baud |= buf[pos++] << 8;
+		cmd->comset.baud |= buf[pos++] << 16;
+		cmd->comset.baud |= buf[pos++] << 24;
+		pd_enqueue_command(pd, cmd);
+		reply->id = REPLY_COM;
+		ret = 0;
+		break;
+	default:
+		break;
+	}
+
+	if (ret != 0 && reply->id == 0) {
+		reply->id = REPLY_NAK;
+		reply->cmd_bytes[0] = OSDP_PD_NAK_RECORD;
+	}
+
+	pd->reply_id = reply->id;
+	if (pd->cmd_id != CMD_POLL) {
+		LOG_DBG(TAG "IN(CMD): 0x%02x[%d] -- OUT(REPLY): 0x%02x",
+		      pd->cmd_id, len, pd->reply_id);
+	}
+}
+
+/**
+ * Returns:
+ * +ve: length of command
+ * -ve: error
+ */
+int pd_build_reply(struct osdp_pd *pd, struct osdp_cmd *reply,
+		   uint8_t *buf, int max_len)
+{
+	uint8_t *smb;
+	int i, data_off, len = 0;
+
+	data_off = phy_packet_get_data_offset(pd, buf);
+	smb = phy_packet_get_smb(pd, buf);
+
+	buf += data_off;
+	max_len -= (data_off + 1); /* +1 for reply->id */
+	if (max_len < 0) {
+		LOG_ERR(TAG "Out of buffer space!");
+		return -1;
+	}
+	buf[len++] = reply->id;
+
+	switch (reply->id) {
+	case REPLY_ACK:
+		break;
+	case REPLY_PDID:
+		if (max_len < 12) {
```

Comment:
Try to avoid magic constants, here and below

Reply:
I really would like to get rid of these magic lengths here. It should not be too big issue to just have something like this in the code:
```
#define MAX_PIID_LEN 12
#define MAX_BAUD_LEN 5
...
```

Reply:
Done.